### PR TITLE
[webserver] CHTTPImageHandler: add GetLastModifiedDate implementation

### PR DESCRIPTION
**Pull request by Rechi**
Posters for movies, tv shows and music can be cached by the browser.

### CODE REVIEW

File: xbmc/network/httprequesthandler/HTTPImageHandler.cpp [hunk 1]
```diff
@@ -36,8 +37,13 @@ CHTTPImageHandler::CHTTPImageHandler(const HTTPRequest &request)
 
     XFILE::CImageFile imageFile;
     const CURL pathToUrl(file);
-    if (imageFile.Exists(pathToUrl))
+    if (imageFile.Exists(pathToUrl)) {
```
Comment (popcornmix) [inline xbmc/network/httprequesthandler/HTTPImageHandler.cpp]: Open bracket should be on a new line

**Comment by Rechi** [comment]
@popcornmix updated

**Comment by MartijnKaijser** [comment]
@Montellese ping

**Comment by Tolriq** [comment]
@Rechi what headers does it update ? 

I have not checked since a long time but there was also an etag problem at some point.

**Comment by Rechi** [comment]
Only ```Last-Modified``` get added to the response header.

**Comment by Tolriq** [comment]
@Rechi Ok no I've tested the problem was not the etag (it's not returned on Windows at least) but the Cache-Control header

Cache-Control: private, max-age=0, no-cache

You need to also have this header to be removed for /image but not the rest.

I've  tested your patch it only adds : Last-Modified: Thu, 19 May 2016 10:12:55 GMT

But leave the other header that will forbid the cache on any client that respect those (meaning nearly all)

File: xbmc/network/httprequesthandler/HTTPImageHandler.h [hunk 2]
```diff
@@ -32,8 +32,14 @@ class CHTTPImageHandler : public CHTTPFileHandler
   virtual IHTTPRequestHandler* Create(const HTTPRequest &request) { return new CHTTPImageHandler(request); }
   virtual bool CanHandleRequest(const HTTPRequest &request);
 
+  virtual bool CanBeCached() const { return true; }
+  virtual bool GetLastModifiedDate(CDateTime &lastModified) const;
```
Comment (Montellese) [inline xbmc/network/httprequesthandler/HTTPImageHandler.h]: Please add the `override` keyword on both methods.

File: xbmc/network/httprequesthandler/HTTPImageHandler.cpp [hunk 3]
```diff
@@ -46,6 +53,15 @@ CHTTPImageHandler::CHTTPImageHandler(const HTTPRequest &request)
   SetFile(file, responseStatus);
 }
 
+bool CHTTPImageHandler::GetLastModifiedDate(CDateTime &lastModified) const
```
Comment (Montellese) [inline xbmc/network/httprequesthandler/HTTPImageHandler.cpp]: The base class `CHTTPFileHandler` already offers helper methods to handle this. You just need to call `SetCanBeCached()` in the constructor and `SetLastModifiedDate()` instead of using your own private `m_lastModified` variable.

File: xbmc/network/httprequesthandler/HTTPImageHandler.cpp [hunk 4]
```diff
@@ -37,7 +38,13 @@ CHTTPImageHandler::CHTTPImageHandler(const HTTPRequest &request)
     XFILE::CImageFile imageFile;
     const CURL pathToUrl(file);
     if (imageFile.Exists(pathToUrl))
+    {
       responseStatus = MHD_HTTP_OK;
+      struct __stat64 buffer;
```
Comment (Montellese) [inline xbmc/network/httprequesthandler/HTTPImageHandler.cpp]: Please see https://github.com/Rechi/xbmc/blob/92cd69039cc531a9b193e1636f1df480d6c76fea/xbmc/network/httprequesthandler/HTTPFileHandler.cpp#L90 for how to do this properly. Maybe we can refactor that logic into a helper method in `CHTTPFileHandler` which can be used by the base and any derived classes.

**Comment by Montellese** [comment]
To fix the issue @Tolriq reported you can either implement/override `GetMaximumAgeForCaching()` or you can manually set the `Cache-Control` header. But I always tried not to manually set any of these "generic" headers and tried to implement it through the `IHTTPRequestHandler` interface. Maybe the `Cache-Control` logic in `CWebServer::FinalizeRequest()` needs some tweaking as well.

**Comment by Tolriq** [comment]
For images correct header should be : 

Cache-Control: public, max-age=XXXXX 

For XXXX the value is debatable, but since there's no etag I guess at least 1 week but @Montellese call.

File: xbmc/network/httprequesthandler/HTTPImageHandler.cpp [hunk 5]
```diff
@@ -37,7 +37,25 @@ CHTTPImageHandler::CHTTPImageHandler(const HTTPRequest &request)
     XFILE::CImageFile imageFile;
     const CURL pathToUrl(file);
     if (imageFile.Exists(pathToUrl))
+    {
       responseStatus = MHD_HTTP_OK;
+      struct __stat64 statBuffer;
+      if (imageFile.Stat(pathToUrl, &statBuffer) == 0)
+      {
+        struct tm *lastChange;
+#ifdef HAVE_LOCALTIME_R
```
Comment (akva2) [inline xbmc/network/httprequesthandler/HTTPImageHandler.cpp]: this will need a probe in the cmake buildsystem if you are to use the define.

Comment (Rechi) [inline xbmc/network/httprequesthandler/HTTPImageHandler.cpp]: that is already used in https://github.com/xbmc/xbmc/blob/master/xbmc/network/httprequesthandler/HTTPFileHandler.cpp#L94

Comment (akva2) [inline xbmc/network/httprequesthandler/HTTPImageHandler.cpp]: sure. that's also broken in that case :)

Comment (Rechi) [inline xbmc/network/httprequesthandler/HTTPImageHandler.cpp]: is [that](https://github.com/xbmc/xbmc/blob/master/project/cmake/scripts/common/ArchSetup.cmake#L96) the probe you want?

Comment (akva2) [inline xbmc/network/httprequesthandler/HTTPImageHandler.cpp]: aha. somebody has already adressed the issue. that's the best fixes ever don't you agree? :)

File: xbmc/network/httprequesthandler/HTTPImageHandler.cpp (hunk 4 above)
Comment (Rechi) [inline xbmc/network/httprequesthandler/HTTPImageHandler.cpp]: What parameter should the helper method have?
maybe ```__stat64``` or ```XFILE::IFile```?

**Comment by Rechi** [comment]
@Montellese is the second commit a good way for the helper method?

**Comment by Montellese** [comment]
Looks good. But you still might have to override `GetMaximumAgeForCaching()` in `CHTTPImageHandler` to get a proper `Cache-Control` header.

**Comment by Rechi** [comment]
@Montellese what will happen if `GetMaximumAgeForCaching()` is set and the picture changes? Are the browsers always checking if `Last-Modified` has changed, or will the old picture be shown until `max-age` expires?

**Comment by Montellese** [comment]
I'd have to re-read the RFC(s) but IIRC `max-age` says that the browser should re-cache the image at least after that period but he's free to do so before that. And if `Last-Modified` changed it should re-cache anyway.

**Comment by Rechi** [comment]
@Montellese have added `GetMaximumAgeForCaching`.
In Firefox the pictures get refreshed if they have changed, also if max-age haven't expired.

File: xbmc/network/httprequesthandler/HTTPImageHandler.h [hunk 6]
```diff
@@ -33,6 +33,7 @@ class CHTTPImageHandler : public CHTTPFileHandler
   virtual bool CanHandleRequest(const HTTPRequest &request);
 
   virtual int GetPriority() const { return 5; }
+  virtual int GetMaximumAgeForCaching() const { return 60 * 60 * 24 * 7; }
```
Comment (Montellese) [inline xbmc/network/httprequesthandler/HTTPImageHandler.h]: Either adjust the commit message or move this into it's own commit.

**Comment by Montellese** [comment]
@Rechi thanks for checking and implementing `GetMaximumAgeForCaching`. This is good to go as soon as my last comment has been resolved.

**Comment by Rechi** [comment]
@Montellese splitted `GetMaximumAgeForCaching` into a separate commit.

**Comment by Montellese** [comment]
Thanks, this is fine with me now. jenkins build this please

**Review by Montellese — APPROVED** [review]
